src/client/sound.h

```diff
@@ -19,75 +19,166 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+	 * Removes finished sounds, steps streamed sounds, and does similar tasks.
+	 * Should not be called while paused.
+	 * @param dtime In seconds.
+	 */
+	virtual void step(f32 dtime) = 0;
+	/**
+	 * Pause all sound playback.
+	 */
+	virtual void pauseAll() = 0;
+	/**
+	 * Resume sound playback after pause.
+	 */
+	virtual void resumeAll() = 0;
 
-class DummySoundManager : public ISoundManager
-{
-public:
-	virtual bool loadSoundFile(const std::string &name, const std::string &filepath)
-	{
-		return true;
-	}
-	virtual bool loadSoundData(const std::string &name, const std::string &filedata)
-	{
-		return true;
-	}
-	void updateListener(const v3f &pos, const v3f &vel, const v3f &at, const v3f &up)
+	/**
+	 * @param pos In node-space.
+	 * @param vel In node-space.
+	 * @param at Vector in node-space pointing forwards.
+	 * @param up Vector in node-space pointing upwards, orthogonal to `at`.
+	 */
+	virtual void updateListener(const v3f &pos, const v3f &vel, const v3f &at,
+			const v3f &up) = 0;
+	virtual void setListenerGain(f32 gain) = 0;
+
+	/**
+	 * Adds a sound to load from a file (only OggVorbis).
+	 * @param name The name of the sound. Must be unique, otherwise call fails.
+	 * @param filepath The path for
+	 * @return true on success, false on failure (ie. sound was already added or
+	 *         file does not exist).
+	 */
+	virtual bool loadSoundFile(const std::string &name, const std::string &filepath) = 0;
+	/**
+	 * Same as `loadSoundFile`, but reads the OggVorbis file from memory.
+	 */
+	virtual bool loadSoundData(const std::string &name, std::string &&filedata) = 0;
+	/**
+	 * Adds sound with name sound_name to group `group_name`. Creates the group
+	 * if non-existent.
+	 * @param sound_name The name of the sound, as used in `loadSoundData`.
+	 * @param group_name The name of the sound group.
+	 */
+	virtual void addSoundToGroup(const std::string &sound_name,
+			const std::string &group_name) = 0;
+
+	/**
+	 * Plays a random sound from a sound group (position-less).
+	 * @param id Id for new sound. Move semantics apply if id > 0.
+	 */
+	virtual void playSound(sound_handle_t id, const SoundSpec &spec) = 0;
+	/**
+	 * Same as `playSound`, but at a position.
+	 * @param pos In node-space.
+	 * @param vel In node-space.
+	 */
+	virtual void playSoundAt(sound_handle_t id, const SoundSpec &spec, const v3f &pos,
+			const v3f &vel) = 0;
+	/**
+	 * Request the sound to be stopped.
+	 * The id should be freed afterwards.
+	 */
+	virtual void stopSound(sound_handle_t sound) = 0;
+	virtual void fadeSound(sound_handle_t sound, f32 step, f32 target_gain) = 0;
+	/**
+	 * Update position and velocity of positional sound.
+	 * @param pos In node-space.
+	 * @param vel In node-space.
+	 */
+	virtual void updateSoundPosVel(sound_handle_t sound, const v3f &pos,
+			const v3f &vel) = 0;
+
+	/**
+	 * Get and reset the list of sounds that were stopped.
+	 */
+	std::vector<sound_handle_t> pollRemovedSounds()
 	{
+		return std::move(m_removed_sounds);
```

Comment:
I'm not 100% on the state of objects after moving, so this would be better
```suggestion
		std::vector<sound_handle_t> ret;
		std::swap(ret, m_removed_sounds);
		return ret;
```

Reply:
According to reference:
>After the move, other is guaranteed to be empty().

(See std::vector constructor.)

Will still do the suggested change, so nobody asks themselves the same question. *shrug*